examples/nlp/ner_transformers.py

```diff
@@ -0,0 +1,326 @@
+            ner_tags = record["ner_tags"]
+            tokens = record["tokens"]
+            f.write(
+                str(len(tokens))
+                + "\t"
+                + "\t".join(tokens)
+                + "\t"
+                + "\t".join(map(str, ner_tags))
+                + "\n"
+            )
+
+
+os.mkdir("data")
+export_to_file("./data/conll_train.txt", conll_data["train"])
+export_to_file("./data/conll_val.txt", conll_data["validation"])
+
+"""
+## Make the NER label lookup table
+"""
+
+"""
+NER labels are usually provided in IOB, IOB2 or IOBES formats. Checkout this link for
+more information:
+[Wikipedia](https://en.wikipedia.org/wiki/Inside%E2%80%93outside%E2%80%93beginning_(tagging))
+[Wikipedia](https://en.wikipedia.org/wiki/Inside%E2%80%93outside%E2%80%93beginning_(tagging))
+
+Note that we start our label numbering from 1 since 0 will be reserved for padding. We
+have a total of 10 labels: 9 from the NER dataset and one for padding
+"""
+
+
+def make_tag_lookup_table():
+    iob_labels = ["B", "I"]
+    ner_labels = ["PER", "ORG", "LOC", "MISC"]
+    all_labels = [(label1, label2) for label2 in ner_labels for label1 in iob_labels]
+    all_labels = ["-".join([a, b]) for a, b in all_labels]
+    all_labels = ["[PAD]", "O"] + all_labels
+    return dict(zip(range(0, len(all_labels) + 1), all_labels))
+
+
+mapping = make_tag_lookup_table()
+mapping
+
+"""
+Get a list of all tokens in the training dataset. This will be used to create the
+vocabulary
+"""
+
+all_tokens = sum(conll_data["train"]["tokens"], [])
+all_tokens_array = np.array(list(map(str.lower, all_tokens)))
+
+counter = Counter(all_tokens_array)
+len(counter)
+
+num_tags = len(mapping)
+vocab_size = 20000
+vocabulary = [token for token, count in counter.most_common(vocab_size - 2)]
+
+# The StringLook class will convert tokens to token IDs
+lookup_layer = tf.keras.layers.experimental.preprocessing.StringLookup(
+    vocabulary=vocabulary
+)
+
+"""
+Create 2 new `Dataset` objects from the training and validation data
+"""
+
+train_data = tf.data.TextLineDataset("./data/conll_train.txt")
+val_data = tf.data.TextLineDataset("./data/conll_val.txt")
+
+# Print out one line to make sure it looks good
+# The first record in the line is the number of tokens. After that we will
+# have all the tokens followed by all the ner tags
+
+list(train_data.take(1).as_numpy_iterator())
+
+"""
+We will be using the following map function to transform the data in the dataset:
+"""
+
+
+def map_func(record):
+    record = tf.strings.split(record, sep="\t")
+    length = tf.strings.to_number(record[0], out_type=tf.int32)
+    tokens = record[1 : length + 1]
+    tokens = tf.strings.lower(tokens)
+    tags = record[length + 1 :]
+    tags = tf.strings.to_number(tags, out_type=tf.int64)
+    tags += 1
+    return tokens, tags
+
+
+batch_size = 32
+train_dataset = (
+    train_data.map(map_func)
+    .map(lambda x, y: (lookup_layer(x), y))
+    .padded_batch(batch_size)
```

Comment:
Why the `padded_batch`?

Reply:
That's because each training example in the dataset has a different length.

Reply:
That would definitely deserve a code comment!

Reply:
Done.